### ISSUE TITLE
Combo box widget values not displayed consistently in different reader software

### ISSUE THREAD

**Issue by leclerce** [pymupdf/PyMuPDF#1311]
## Description
When filling a pdf template with combo box type widgets, not all PDF reader software correctly displays the value set in fitz' field_value attribute (or any other python library setting \V and/or \DV in a combo box widget)

Readers that properly display the field_value (not an exhaustive list):
Linux: Evince, Firefox, Brave
Windows: Firefox, Chrome, Edge
MacOS: Preview, Foxit Reader, Chrome, Safari

Readers that do not display the field_value:
Windows: Adobe Acrobat Pro 2020
MacOS: Adobe Acrobat Reader DC 2020

Analysis of a filled form using the "Prepare Form" tool in Acrobat pro shows that none of the values are detected as set (Value, Default Value and Exported Value).  Unfortunately this tool only exists on the Pro version of Acrobat.

## To Reproduce

```
pdf_template = fitz.open("/path/to/template.pdf")
data_dict = {
    'Currency': 'AUD - Australian Dollarlydoo'
}

#example for a widget with field_name Currency with choice_values array 
# ['Select', 'AUD - Australian Dollarydoo', 'USD - US Dollar']

for page in pdf_template.pages():
    for widget in page.widgets():
        if widget.field_type == fitz.PDF_WIDGET_TYPE_COMBOBOX:
            if data_dict.get(widget.field_name) in widget.choice_values:
                widget.field_value = data_dict.get(widget.field_name, "Select")
                widget.update()

pdf_template.save("/path/to/output.pdf")
```

## Expected behavior 
Widget value is consistently displayed

## Your configuration 
Python 3.8.12 Linux
PyMuPDF 1.18.19 Version date 2021-09-16 16:45:29 Built for Python 3.8 on linux 64-bit

## Additional context 
I realize this is an issue with the reader software and the PDF document spec itself more so than this library, however I thought someone here may have come across this problem before. 

From the PDF 1.7 spec pp.445:
https://www.levien.com/pdf/PDF32000_2008.pdf

"The Opt array specifies the list of options in the choice field, each of which shall be represented by a text string that shall be displayed on the screen.  Each element of the Opt array contains either the text string by itself or a two-element array, whose second element is the text string and whose first element is a text string representing the export value that shall be used when exporting interactive form field data from the document"

This seems to indicate that the choice_values array may need to be converted.

**Comment by JorjMcKie** [comment]
As per the `/Opt` array, there was another error in my code: I assumed it to be a Python list (which it need not be - tuples are allowed as per documentation). So if a tuple of choice values is being handed in, my code ignores it ...

**Comment by leclerce** [comment]
I can confirm that the field flag bits are set as expected.  I am not actually convinced that the widget needs to be set to editable if you expect to use an existing choice_value in the drop down, the way I'm reading that spec is bit 19 is set if you want to add additional values to the existing list.

Setting the field_value of the widget to one of the existing choice_value entries does not display in Adobe though.  :(


```
If set, the combo box shall include an editable text box as well
as a drop-down list; if clear, it shall include only a drop-down
list. This flag shall be used only if the Combo flag is set.
```

**Comment by JorjMcKie** [comment]
>  I am not actually convinced that the widget needs to be set to editable if you expect to use an existing choice_value

No, it is not. This was just to demonstrate that additional prperties must be OR into that field, while respecting any bit positions that take part in the field type as such.

As per the Adobe remaining issue, I believe this is because Acrobat additionally sets a `/I` key in the annot definition, with the value of the index of the chosen entry in the list. At least I have seen Acrobat setting this value ... and I do not currently modify it.
Maybe it works if just removing that key / value ... it is optional except when each choice item consists of two subitems (one for display, the 2nd one as the actual value).

If you could try to "manually" remove and see if that solves the issue, I can include htis in the logic:
```python
doc.xref_set_key(annot.xref, "I", "null")
```

**Comment by JorjMcKie** [comment]
Thansk for trying. We are on the same page.
I can confirm that my assumption is true:
```python
>>> doc=fitz.open("x.pdf")
>>> page=doc[0]
>>> for w in page.widgets():
	print(w.xref, w.field_value)

	
19 Times-Roman
21 Yes
23 Frankfurt
25 Dänemark
...
>>> for w in page.widgets():
	if w.xref==25: break

	
>>> w
'ComboBox' widget on page 0 of x.pdf
>>> w.field_value
'Dänemark'
>>> w.field_value=w.choice_values[-2]
>>> w.update()
Portugal
>>> print(doc.xref_object(25))
<<
  /AP <<
    /N 12 0 R
  >>
  /BS <<
    /S /S
    /W 1
  >>
  /DA (0 0 1 rg /Helv 11 Tf)
  /F 4
  /FT /Ch
  /Ff 33947648
  /I [ 3 ]  % <== still the value set by Adobe (Dänemark)!
  /MK <<
    /BG [ 1 1 0 ]
  >>
  /NM (fitz-W3)
  /Opt [ (Spanien) (Frankreich) (Holland) <FEFF004400E4006E0065006D00610072006B>
      (Schweden) (Norwegen) (England) (Polen) (Russland) (Italien)
      (Portugal) (Griechenland) ]
  /P 7 0 R
  /Rect [ 202 637 352 652 ]
  /Subtype /Widget
  /T (ComboBox-1)
  /TU (an editable combo box ...)
  /Type /Annot
  /V (Portugal)
>>
>>> doc.xref_set_key(25, "I", "null")  # remove the /I value
>>> print(doc.xref_object(25))  # confirm it worked
<<
  /AP <<
    /N 12 0 R
  >>
  /BS <<
    /S /S
    /W 1
  >>
  /DA (0 0 1 rg /Helv 11 Tf)
  /F 4
  /FT /Ch
  /Ff 33947648
  /I null
  /MK <<
    /BG [ 1 1 0 ]
  >>
  /NM (fitz-W3)
  /Opt [ (Spanien) (Frankreich) (Holland) <FEFF004400E4006E0065006D00610072006B>
      (Schweden) (Norwegen) (England) (Polen) (Russland) (Italien)
      (Portugal) (Griechenland) ]
  /P 7 0 R
  /Rect [ 202 637 352 652 ]
  /Subtype /Widget
  /T (ComboBox-1)
  /TU (an editable combo box ...)
  /Type /Annot
  /V (Portugal)
>>
>>> doc
Document('x.pdf')
>>> doc.ez_save("y.pdf",pretty=True)  # save new file
```
Displaying the new file under Adobe Acobat shows "Portugal" as intended.
So I will now remove the ``/I`` key when updating list and combo boxes.

**Comment by leclerce** [comment]
That would be great, thanks!

**Comment by JorjMcKie** [comment]
I have further thought about this:
* The `/I` entry is mostly optional.
* If it specifies a different choice item, than that in the `/V` field, then `/V` must be taken. So, Adobe and other viewers act **_contrary_** to that specification!
* The `/I` is **_required only_** if (1) the choice items consist of two subitems each, and (2) there exist two or more items having the same **export** subitem (but different **name** subitems).

So I am wondering if I am being too radical when I always simply remove the `/I`. Maybe I should add a new ``Widget``  attribute - ``choice_indices`` - that allows selecting one or more of the items. If not ``None`` and in the valid range, I would take the identified choice value(s) and (1) use it to fill `/V` and (2) insert / overwrite `/I` with that index array.

What is your reaction?

**Comment by leclerce** [comment]
For my current use case, it does not appear that my choice items have the (value, export value) arrangement so removing `/I` should work for all of the combo boxes on my form.  

I would also agree that Adobe doesn't seem to follow their own spec for this widget type.  None of the other viewers I tried had this issue!

As a general solution conforming to the PDF document spec you're correct though - it is part of the spec to allow a combo box with multi-select, and to have the choice items defined as an array (value, export value) so a `choice_indices` attribute may prove to be useful for that purpose, or at the very least as a workaround to get uncooperative viewers to display what you'd like them to display.

Quoting the relevant sections of the spec here for reference:

Regarding the **/I** key:
```
(Sometimes required, otherwise optional; PDF 1.4) For choice fields that allow
multiple selection (MultiSelect flag set), an array of integers, sorted in ascending
order, representing the zero-based indices in the Opt array of the currently
selected option items. This entry shall be used when two or more elements in the
Opt array have different names but the same export value or when the value of
the choice field is an array. This entry should not be used for choice fields that do
not allow multiple selection. If the items identified by this entry differ from those in
the V entry of the field dictionary (see discussion following this Table), the V entry
shall be used.
```

Regarding **/Opt** and **/V** w.r.t list and combo box widgets:
```
The Opt array specifies the list of options in the choice field, each of which shall be represented by a text string
that shall be displayed on the screen. Each element of the Opt array contains either this text string by itself or a
two-element array, whose second element is the text string and whose first element is a text string representing
the export value that shall be used when exporting interactive form field data from the document.

The field dictionary’s V (value) entry identifies the item or items currently selected in the choice field. If the field does not allow multiple selection—that is, if the MultiSelect flag (PDF 1.4) is not set—or if multiple selection is supported but only one item is currently selected, V is a text string representing the name of the selected item, as given in the field dictionary’s Opt array. If multiple items are selected, V is an array of
such strings. (For items represented in the Opt array by a two-element array, the name string is the second of the two array elements.) The default value of V is null, indicating that no item is currently selected.
```

**Comment by JorjMcKie** [comment]
Resolved in new version 1.19.0 which is currently being uploaded.
